spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/admin/CachingComposingSupplier.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.core.admin;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A supplier of objects that relies on another supplier and caches
+ * provided results.
+ *
+ * @param <T> the type of objects this supplier produces.
+ * @param <U> the type of objects this supplier bases its products on.
+ * @author Chengyuan Zhao
+ */
+public class CachingComposingSupplier<T, U> implements Supplier<T> {
```

Comment:
Make package-private? Looks like an implementation detail.

Reply:
This is used in the auto-config module, which is a different package